src/MatrixFields/MatrixFields.jl

```diff
@@ -19,46 +19,60 @@ for them:
 - Integration with `RecursiveApply`, e.g., the entries of `matrix_field` can be
     `Tuple`s or `NamedTuple`s instead of single values, which allows
     `matrix_field` to represent multiple band matrices at the same time
+- Integration with `Operators`, e.g., the `matrix_field` that gets applied to
+    the argument of any `FiniteDifferenceOperator` `op` can be obtained using
+    the `FiniteDifferenceOperator` `operator_matrix(op)`
 - Conversions to native array types, e.g., `field2arrays(matrix_field)` can
     convert each column of `matrix_field` into a `BandedMatrix` from
     `BandedMatrices.jl`
 - Custom printing, e.g., `matrix_field` gets displayed as a `BandedMatrix`,
     specifically, as the `BandedMatrix` that corresponds to its first column
+
+This module also adds support for defining and manipulating sparse block
+matrices of `Field`s. Specifically, it adds the `FieldMatrix` type, which is a
+dictionary that maps pairs of `FieldName`s to `ColumnwiseBandMatrixField`s or
+multiples of `LinearAlgebra.I`. This comes with the following functionality:
+- Addition and subtraction, e.g., `@. field_matrix1 + field_matrix2`
+- Matrix-vector multiplication, e.g., `@. field_matrix * field_vector`
+- Matrix-matrix multiplication, e.g., `@. field_matrix1 * field_matrix2`
+- Integration with `RecursiveApply`, e.g., the entries of `field_matrix` can be
+    specified either as matrix `Field`s of `Tuple`s or `NamedTuple`s, or as
+    separate matrix `Field`s of single values
+- The ability to solve linear equations using `FieldMatrixSolver`, which is a
+    generalization of `ldiv!` that is designed to optimize solver performance
 """
 module MatrixFields
 
-import CUDA: @allowscalar
-import LinearAlgebra: UniformScaling, Adjoint, AdjointAbsVec
+import CUDA
+import LinearAlgebra: I, UniformScaling, Adjoint, AdjointAbsVec, mul!, inv
 import StaticArrays: SMatrix, SVector
 import BandedMatrices: BandedMatrix, band, _BandedMatrix
+import ClimaComms
 
 import ..Utilities: PlusHalf, half
 import ..RecursiveApply:
     rmap, rmaptype, rpromote_type, rzero, rconvert, radd, rsub, rmul, rdiv
+import ..RecursiveApply: ⊠, ⊞, ⊟
 import ..DataLayouts: AbstractData
 import ..Geometry
 import ..Spaces
 import ..Fields
 import ..Operators
 
-export ⋅
 export DiagonalMatrixRow,
     BidiagonalMatrixRow,
     TridiagonalMatrixRow,
     QuaddiagonalMatrixRow,
     PentadiagonalMatrixRow
+export FieldVectorKeys, FieldVectorView, FieldVectorViewBroadcasted
+export FieldMatrixKeys, FieldMatrix, FieldMatrixBroadcasted
+export @name, ⋅, FieldMatrixSolver, field_matrix_solve!
```

Comment:
`@name` is _very_ generic. I don't think this should be exported.

Reply:
Ok, I'll change it!

Reply:
(Or `@fieldname`?)

Reply:
Idk, I'm inclined to say no because it looks an awful lot like an individual macro for something like `fieldnames`. The entire field names part of this code looks like it could/should be a part of a separate package. The API is not clear, and there are a lot of methods (this is perhaps a Julia issue, though). Maybe @simonbyrne can chime in here?

Reply:
It's essentially a way to store nested `propertynames`; i.e., a type-stable version of `propertychains`.

Reply:
It will probably be useful to move the `FieldName`s infrastructure to ClimaUtilities in the future. For now, though, we might want to leave it in here to sort out the interface issues more easily. @charleskawczynski Can I just leave things as they are for now, but with `@name` unexported? That would let people start testing new Jacobians in ClimaAtmos while I make further interface changes in other PRs.

Reply:
Is there anything I can improve in the docs to make the API more clear?

Reply:
> Can I just leave things as they are for now, but with @name unexported?

I think that's fine